File: javascript/bst/src/lib/bst.js

class Node {
    constructor() {
	this.data = 0
	this.leftChild = null
	this.rightChild = null
    }

    insert(v) {

	let currNode = this
	
	if (v < currNode.data) {
	    if (currNode.leftChild == null) {
		let n = new Node()
		n.data = v
		currNode.leftChild = n
	    } else {
		currNode.leftChild.insert(v)
	    }
	} else {
	    if (currNode.rightChild == null) {
		let n = new Node()
		n.data = v
		currNode.rightChild = n
	    } else {
		currNode.rightChild.insert(v)
	    }
	}
    }
    
    getItems() {

	let vals = []

	let currNode = this
	
	if (currNode.leftChild != null) {
	    vals.push(...currNode.leftChild.getItems())
	}

	vals.push(currNode.data)

	if (currNode.rightChild != null) {
	    vals.push(...currNode.rightChild.getItems())
	}

	return vals
    }
}

class BST  {
    contructor() {
 	this.root = new Node()
    }
    
    insert(v) {

	// 0 not allowed
	if (v == 0)
	    return
	
	// TODO check if v already exists
		
	if (this.root == null) {
	    this.root = new Node()
	    this.root.data = v
	    return
	}
	
	this.root.insert(v)
    }

    getItems() {

	if (this.root == null)
	    return []
	
	//console.log(this.root)
	let vals = this.root.getItems()

	return vals
    }

    search(v) {

	let currNode = this.root

	while (currNode != null) {
	    if (currNode.data == v) {
		return true
	    } else if (v < currNode.data) {
		currNode = currNode.leftChild
	    } else {
		currNode = currNode.rightChild
	    }
	}

	return false
    }
	    
}

export default BST

File: python/bst/bst.py

class Node:
    left = None
    right = None
    val = None
    
class BST():

    def __init__(self):
        self.root_node = None

    def add_node(self, v):
        if self.root_node == None:
            n = Node();
            n.val = v
            self.root_node = n
        else:
            self._add_node(self.root_node, v)

    def _add_node(self, node, v):
        if v <= node.val:
            if node.left == None:
                n = Node()
                n.val = v
                node.left = n
            else:
                self._add_node(node.left, v)
        else:
            if node.right == None:
                n = Node()
                n.val = v
                node.right = n
            else:
                self._add_node(node.right, v)

    def get_nodes(self):
        if self.root_node != None:
            nodes = []
            return self._get_nodes(self.root_node, nodes)
        else:
            return []

    def _get_nodes(self, node, nodes):
        
        if node.left != None:
            self._get_nodes(node.left, nodes)

        nodes.append(node.val)

        if node.right != None:
            self._get_nodes(node.right, nodes)

        return nodes

    def search(self, val):

        curr_node = self.root_node

        while curr_node != None:
            if curr_node.val == val:
                return True
            elif val <= curr_node.val:
                curr_node = curr_node.left
            else:
                curr_node = curr_node.right

        return False
